### DISCUSSION TITLE
Specify Enum loading and dumping?

### DISCUSSION THREAD

**Discussion by CaptainDriftwood** [wyfo/apischema#161] [Q&A] [answered]
Sorry if the title is slightly misleading. Basically wondering if there is a way to do what marshmallow_enum does?
https://pypi.org/project/marshmallow-enum/

Is there a way to specify if one wants to use the enum value for loading/dumping, or the enum type?

**Comment by wyfo** [comment, marked as answer]
As described in the [documentation](https://wyfo.github.io/apischema/data_model/#enumeration), *apischema* (de)serializes enumeration based on the `Enum` member **value**, not the name.

However, it is possible to modify this behavior by registering a [conversion](https://wyfo.github.io/apischema/conversions/). For example, to use `Enum` names instead of values, you could write the following:
```python
from enum import Enum
from typing import Literal, TypeVar

import apischema

EnumCls = TypeVar("EnumCls", bound=type[Enum])


def as_names(enum_cls: EnumCls) -> EnumCls:
    names = Literal[tuple(elt.name for elt in enum_cls)]  # type: ignore
    apischema.deserializer(
        apischema.conversions.Conversion(
            lambda name: getattr(enum_cls, name), source=names, target=enum_cls
        )
    )
    apischema.serializer(
        apischema.conversions.Conversion(
            lambda elt: elt.name, source=enum_cls, target=names
        )
    )
    return enum_cls

@as_names
class MyEnum(Enum):
    FOO = object()
    BAR = object()

assert apischema.deserialize(MyEnum, "FOO") == MyEnum.FOO
assert apischema.serialize(MyEnum, MyEnum.FOO) == "FOO"
```

By the way, this `as_names` decorator has been added to the library in #160; that's fun because I've merged it a few hours before your question. It will be available (with other cool new features) in the next release of *apischema*, normally before the end of the month.
